python/ray/_raylet.pyx

```diff
@@ -981,17 +940,145 @@ cdef execute_task(
                         "See https://github.com/ray-project/ray/issues/28689.")
 
+
+cdef execute_task_with_cancellation_handler(
+        const CAddress &caller_address,
+        CTaskType task_type,
+        const c_string name,
+        const CRayFunction &ray_function,
+        const unordered_map[c_string, double] &c_resources,
+        const c_vector[shared_ptr[CRayObject]] &c_args,
+        const c_vector[CObjectReference] &c_arg_refs,
+        const c_string debugger_breakpoint,
+        const c_string serialized_retry_exception_allowlist,
+        c_vector[c_pair[CObjectID, shared_ptr[CRayObject]]] *returns,
+        c_vector[c_pair[CObjectID, shared_ptr[CRayObject]]] *dynamic_returns,
+        c_bool *is_retryable_error,
+        c_bool *is_application_error,
+        # This parameter is only used for actor creation task to define
+        # the concurrency groups of this actor.
+        const c_vector[CConcurrencyGroup] &c_defined_concurrency_groups,
+        const c_string c_name_of_concurrency_group_to_execute,
+        c_bool is_reattempt):
+
+    is_application_error[0] = False
+    is_retryable_error[0] = False
+
+    worker = ray._private.worker.global_worker
+    manager = worker.function_actor_manager
+    cdef:
+        dict execution_infos = manager.execution_infos
+        CoreWorker core_worker = worker.core_worker
+        JobID job_id = core_worker.get_current_job_id()
+        TaskID task_id = core_worker.get_current_task_id()
+        CFiberEvent task_done_event
+        c_vector[shared_ptr[CRayObject]] dynamic_return_ptrs
+
+    task_name = name.decode("utf-8")
+    title = f"ray::{task_name}()"
+
+    # Automatically restrict the GPUs available to this task.
+    ray._private.utils.set_cuda_visible_devices(ray.get_gpu_ids())
+
+    # Initialize the actor if this is an actor creation task. We do this here
+    # before setting the current task ID so that we can get the execution info,
+    # in case executing the main task throws an exception.
+    function_descriptor = CFunctionDescriptorToPython(
+        ray_function.GetFunctionDescriptor())
+    if <int>task_type == <int>TASK_TYPE_ACTOR_CREATION_TASK:
+        actor_class = manager.load_actor_class(job_id, function_descriptor)
+        actor_id = core_worker.get_actor_id()
+        actor = actor_class.__new__(actor_class)
+        worker.actors[actor_id] = actor
+        # Record the actor class via :actor_name: magic token in the log.
+        #
+        # (Phase 1): this covers code run before __init__ finishes.
+        # We need to handle this separately because `__repr__` may not be
+        # runnable until after `__init__` (e.g., if it accesses fields
+        # defined in the constructor).
+        actor_magic_token = "{}{}\n".format(
+            ray_constants.LOG_PREFIX_ACTOR_NAME, actor_class.__name__)
+        # Flush to both .out and .err
+        print(actor_magic_token, end="")
```

Comment:
@stephanie-wang I've seen the `KeyboardInterrupt` for task cancellation raised during this block, at this line in particular (`print` will cause the Python interpreter to check for signals). Would it be possible to have this covered?

Reply:
I moved the `current_task_id` set to happen after this code, so I don't think this is possible anymore.